core/src/main/scala/org/apache/spark/scheduler/BlacklistStrategy.scala

```diff
@@ -0,0 +1,182 @@
+      executorIdToFailureStatus: mutable.HashMap[String, FailureStatus],
+      atomTask: StageAndPartition,
+      clock: Clock): Set[String] = {
+    getExecutorBlacklistInfo(executorIdToFailureStatus).keys.toSet
+  }
+
+  def getNodeBlacklist(
+      executorIdToFailureStatus: mutable.HashMap[String, FailureStatus],
+      clock: Clock): Set[String] = {
+    getExecutorBlacklistInfo(executorIdToFailureStatus)
+      .groupBy{case (id, failureStatus) => failureStatus.host}
+      .filter {case (host, executorIdToFailureStatus) =>
+        executorIdToFailureStatus.size > maxBlacklistedExecutors}
+      .keys.toSet
+  }
+}
+
+/**
+ * This strategy is applied as default to keep the same semantics as original. It's an task
+ * related strategy. If an executor failed running "task A", then we think this executor is
+ * blacked for "task A". And we think the executor is still healthy for other task. node blacklist
+ * is always empty.
+ *
+ * It was the standard behavior before spark 1.6
+ */
+private[scheduler] class SingleTaskStrategy(
+    val expireTimeInMilliseconds: Long) extends BlacklistStrategy {
+  def getExecutorBlacklist(
+      executorIdToFailureStatus: mutable.HashMap[String, FailureStatus],
+      atomTask: StageAndPartition,
+      clock: Clock): Set[String] = {
+    executorIdToFailureStatus.filter{
+      case (_, failureStatus) => failureStatus.numFailuresPerTask.keySet.contains(atomTask) &&
+        clock.getTimeMillis() - failureStatus.updatedTime < expireTimeInMilliseconds
+    }.keys.toSet
+  }
+
+  def getNodeBlacklist(
+      executorIdToFailureStatus: mutable.HashMap[String, FailureStatus],
+      clock: Clock): Set[String] = Set.empty[String]
+}
+
+/**
+ * Support getNodeBlacklistForStage. With this strategy, once executor failed running a task, we
+ * put all executors on the same node into blacklist, so all tasks on the same stage will not be
+ * allocated to that node.
+ */
+private[scheduler] class AdvancedSingleTaskStrategy(
+    expireTimeInMilliseconds: Long) extends SingleTaskStrategy(expireTimeInMilliseconds) {
+
+  override def getNodeBlacklistForStage(
+      executorIdToFailureStatus: mutable.HashMap[String, FailureStatus],
+      stageId: Int,
+      clock: Clock): Set[String] = {
+    executorIdToFailureStatus.filter{
+      case (_, failureStatus) =>
+        failureStatus.numFailuresPerTask.keySet.map(_.stageId).contains(stageId) &&
+        clock.getTimeMillis() - failureStatus.updatedTime < expireTimeInMilliseconds
+    }.values.map(_.host).toSet
+  }
+}
+
+/**
+ * Create BlacklistStrategy instance according to SparkConf
+ */
+private[scheduler] object BlacklistStrategy {
+  def apply(sparkConf: SparkConf): BlacklistStrategy = {
+    val timeout = sparkConf.getTimeAsMs("spark.scheduler.blacklist.timeout",
+        sparkConf.getLong("spark.scheduler.executorTaskBlacklistTime", 0L).toString() + "ms")
+    sparkConf.get("spark.scheduler.blacklist.strategy", "singleTask") match {
```

Comment:
As we talked offline, I am wondering if we can simplify the configuration by just enabling the experimental feature or not?